src/dbnode/storage/shard.go

```diff
@@ -2725,77 +2727,92 @@ func (s *dbShard) AggregateTiles(
 	}
 
 	nextVolume := latestTargetVolume + 1
-	writer, err := fs.NewStreamingWriter(
-		fs.StreamingWriterOptions{
-			NamespaceID:         s.namespace.ID(),
-			ShardID:             s.ID(),
-			Options:             s.opts.CommitLogOptions().FilesystemOptions(),
-			BlockStart:          opts.Start,
-			BlockSize:           s.namespace.Options().RetentionOptions().BlockSize(),
-			VolumeIndex:         nextVolume,
-			PlannedRecordsCount: uint(maxEntries),
-		},
-	)
-	if err != nil {
-		return 0, err
-	}
-
-	if err := writer.Open(); err != nil {
+	writerOpenOpts := fs.StreamingWriterOpenOptions{
+		NamespaceID:         s.namespace.ID(),
+		ShardID:             s.ID(),
+		BlockStart:          opts.Start,
+		BlockSize:           s.namespace.Options().RetentionOptions().BlockSize(),
+		VolumeIndex:         nextVolume,
+		PlannedRecordsCount: uint(maxEntries),
+	}
+	if err = writer.Open(writerOpenOpts); err != nil {
 		return 0, err
 	}
 
 	var (
 		processedTileCount atomic.Int64
 		multiErr           xerrors.MultiError
-		dp                 ts.Datapoint
+		annotationPayload  annotation.Payload
+		downsampledValues  = make([]tile.DownsampledValue, 0, 4)
 		segmentCapacity    int
 		writerData         = make([][]byte, 2)
 	)
 
-READER:
 	for readerIter.Next() {
-		seriesIter, id, encodedTags := readerIter.Current()
+		var (
+			seriesIter, id, encodedTags = readerIter.Current()
+			prevFrameLastValue          = math.NaN()
+			hasData                     bool
+			handleValueResets           bool
+			firstAnnotation             ts.Annotation
+			writeErr                    error
+		)
 
 		for seriesIter.Next() {
 			frame := seriesIter.Current()
-			unit, singleUnit := frame.Units().SingleValue()
-			annotation, singleAnnotation := frame.Annotations().SingleValue()
 
 			if frameValues := frame.Values(); len(frameValues) > 0 {
```

Comment:
This conditional should be flipped to reduce nesting (breaks are harder to grok when nesting gets this deep)

Reply:
👍